### PR TITLE
Documentation update

### PR DESCRIPTION
**Pull request by ab-pm**
As promised. More to follow.

### CODE REVIEW

File: src/pages/graphile-build/hooks.md [hunk 1]
```diff
@@ -20,54 +21,42 @@ const MyType = newWithHooks(GraphQLObjectType, spec);
...
 ```
 
-Hooks are registered via a call to `builder.hook(hookName, hookFunction)`.
+Hooks are registered via a call to [`builder.hook(hookName, hookFunction)`](/graphile-build/schema-builder/#hookhookname-hookfunction).
 
 Every `hookFunction` must synchronously return a value - either the value that it was passed
 or a derivative of it (preferably immutable, but we're not strict on that).
 
-### Which hook to attach to: `hookName`
+Hook functions for one hook by default run in the order they were registered, that is why the order of plugins is sometimes relevant.
```
Comment (ab-pm) [inline src/pages/graphile-build/hooks.md]: See https://github.com/graphile/graphile-engine/pull/416 - this might require an update. Probably the full description should reside at the `/graphile-build/schema-builder/#hookhookname-hookfunction`.

File: src/components/Layout/index.scss [hunk 2]
```diff
@@ -456,6 +456,10 @@ ol {
   }
 }
 
+a code {
+  @extend .underline;
+}
+
```
Comment (benjie) [inline src/components/Layout/index.scss]: Please move this to a separate PR; we can get it merged independently then :+1:

Comment (ab-pm) [reply src/components/Layout/index.scss]: I think there weren't any links on code yet (or did I just not see them?), so this was really for the parts I've written.

File: src/pages/graphile-build/hooks.md [hunk 3]
```diff
@@ -20,54 +21,42 @@ const MyType = newWithHooks(GraphQLObjectType, spec);
...
+3.  The `build` hook allows plugins to add new utility methods to the `build` object itself, or overwrite previously declared ones.
+4.  The `init` hook is the setup phase where the schema types get created by the plugins,
+    using the [`newWithHooks`](/graphile-build/build-object/#newwithhookstype-spec-scope)
+    or [`addType`](/graphile-build/build-object/#addtypetype-graphqlnamedtype) methods of the build object.
+    For example, the [default `StandardTypesPlugin`](/graphile-build/default-plugins/#standardtypesplugin) registers the builtin scalars,
+    or the PostGraphile plugins register the types found in the introspection results.
+5.  The schema is constructed using `newWithHooks(GraphQLSchema, …)`, where the `query`, `mutation` and `subscription` root operations
+    are configured by the respective [default plugins](/graphile-build/default-plugins)
+    and [other schema options](https://github.com/graphql/graphql-js/blob/v14.5.6/src/type/schema.js#L318-L324) can be adjusted.
+6.  The `finalize` hook allows plugins to alter the schema instance and all its contents.
```
Comment (benjie) [inline src/pages/graphile-build/hooks.md]:
```suggestion
6.  The `finalize` hook allows plugins to replace the schema that has been built with an alternative (likely derivative) schema, should that be desired. It also opens an opportunity to do something with the built schema (for example log it out) before it is returned.
```

File: src/pages/graphile-build/hooks.md [hunk 4]
```diff
@@ -20,54 +21,42 @@ const MyType = newWithHooks(GraphQLObjectType, spec);
...
 
-- When creating a `GraphQLObjectType` via
-  `newWithHooks`:
+During the `init` hook, the [hooks for the respective GraphQL types](/graphile-build/all-hooks/) are applied [in the
+`newWithHooks` function](https://github.com/graphile/graphile-engine/blob/v4.4.4/packages/graphile-build/src/makeNewBuild.js#L329),
```
Comment (benjie) [inline src/pages/graphile-build/hooks.md]: I think footnotes might be more appropriate for this level of detail - I don't think people starting with the library necessarily want links to the source.

File: src/pages/graphile-build/hooks.md [hunk 5]
```diff
@@ -115,7 +98,7 @@ used ones:
 - `scope` - an object based on the third argument to `newWithHooks` or
   `fieldWithHooks`; for deeper hooks (such as `GraphQLObjectType:fields:field`)
   the scope from shallower hooks (such as `GraphQLObjectType`) are merged in.
-- `Self` - only available on hooks that are called after the object is created
+- `Self` - only available on deferred hooks that are called after the object is created
```
Comment (benjie) [inline src/pages/graphile-build/hooks.md]: This sentence is now a little redundant, lets move the redundancy to parenthesis:

```suggestion
- `Self` - only available on deferred hooks (those that are called after the object is created,
   e.g. `GraphQLOb...
```

File: src/pages/graphile-build/plugin-options.md [hunk 6]
```diff
@@ -6,9 +6,8 @@ title: Plugin Options
 
 ## Plugin Options
 
-The second argument to `buildSchema` is [the
-options](/graphile-build/plugin-options/) which are made available to every
-plugin (as their second argument).
+The second argument to the [`graphile-build` functions](http://localhost:8000/graphile-build/graphile-build/#the-graphile-build-module)
```
Comment (benjie) [inline src/pages/graphile-build/plugin-options.md]:
```suggestion
The second argument to [`buildSchema` and `getBuilder`](http://localhost:8000/graphile-build/graphile-build/#the-graphile-build-module)
```

File: src/pages/graphile-build/plugin-options.md [hunk 7]
```diff
@@ -6,9 +6,8 @@ title: Plugin Options
...
+is an **`options`** object with arbitrary propreties which is made available to every plugin as its second argument.
```
Comment (benjie) [inline src/pages/graphile-build/plugin-options.md]:
```suggestion
is an **`options`** object with arbitrary properties which is made available to every plugin as its second argument.
```

I think it's also included on the `build` object now? I have a feeling that this turned out to be a better approach, but I don't remember the details off-hand.

File: src/pages/graphile-build/all-hooks.md [hunk 8]
```diff
@@ -10,54 +10,68 @@ The following hooks are currently supported, but more may be added in future.
...
-  - `GraphQLObjectType:interfaces` to add additional interfaces to this object type
-  - `GraphQLObjectType:fields` to add additional fields to this object type (is
-    ran asynchronously and gets a reference to the final GraphQL Object as
-    `Self` in the context)
-  - `GraphQLObjectType:fields:field`: to add any root-level attributes to an
+  `newWithHooks`, we'll execute the following hooks:
+
+  - `GraphQLObjectType` to add/remove any root-level attributes, e.g. a description
+  - `GraphQLObjectType:interfaces` to add additional interfaces to this object type or remove them
+  - `GraphQLObjectType:fields` (*deferred*) to add additional fields to this object type or remove them.
```
[the remark this answers is not in the document]
Comment (ab-pm) [reply src/pages/graphile-build/all-hooks.md]: I just kept it - removing is currently mentioned at https://www.graphile.org/graphile-build/hooks/#which-hook-to-attach-to-hookname. Also I think the graphile-engine docs should mention this separately from postgraphile/extending-raw - but it might be a good idea to add the warning about adding things back, including an explanation of how lookahead is affected (I'm not sure I understand this).

File: src/pages/graphile-build/all-hooks.md [hunk 9]
```diff
@@ -10,54 +10,68 @@ The following hooks are currently supported, but more may be added in future.
...
-  - `GraphQLEnumType:values:value` to customize an individual value from above
+  - `GraphQLEnumType` add/remove any root-level attributes, e.g. add a description
+  - `GraphQLEnumType:values` add/remove values
+  - `GraphQLEnumType:values:value` customize an individual value from above
+
+- `GraphQLSchema`: This event defines the root-level schema; hook it to add `query`,
+  `mutation`, `subscription` root operations or similar schema configurations.
+
+- `finalize`: This event is triggered when the schema has been constructed,
+  hook it to modify or wrap the schema instance.
```
[the remark this answers is not in the document]
Comment (ab-pm) [reply src/pages/graphile-build/all-hooks.md]: Adding `built` sounds like a good idea, the comment seems rather repetitive. Also the difference should become clear when looking at the input type. I thought about merging those sections, having both input type and explanation per hook?

File: src/pages/graphile-build/all-hooks.md [hunk 10]
```diff
@@ -10,54 +10,68 @@ The following hooks are currently supported, but more may be added in future.
...
+asynchronous.
 
 ### Input types
 
 Depending on the hook being called the input object might be an array (as in
 the case of `GraphQLObjectType:interfaces`) or an object (as in all other
 cases, currently). More specifically, the types for each hook are:
 
-- build - a `Build` object (see below)
+- inflection - a plain object with some core inflection methods, built in [`makeNewBuild`](https://github.com/graphile/graphile-engine/blob/v4.4.4/packages/graphile-build/src/makeNewBuild.js#L929-L997)
```
[the remark this answers is not in the document]
Comment (ab-pm) [reply src/pages/graphile-build/all-hooks.md]: Awww, I deliberately went for tag references :-) They make for shorter URLs, and you can see much easier how outdated the source reference is. Did you ever move a tag involving large code changes?

File: src/pages/graphile-build/hooks.md [hunk 11]
```diff
@@ -20,54 +21,42 @@ const MyType = newWithHooks(GraphQLObjectType, spec);
...
+2.  The `inflection` hook allows plugins to add new inflection methods to the `build.inflection` object, or overwrite previously declared ones.
+3.  The `build` hook allows plugins to add new utility methods to the `build` object itself, or overwrite previously declared ones.
+4.  The `init` hook is the setup phase where the schema types get created by the plugins,
+    using the [`newWithHooks`](/graphile-build/build-object/#newwithhookstype-spec-scope)
+    or [`addType`](/graphile-build/build-object/#addtypetype-graphqlnamedtype) methods of the build object.
+    For example, the [default `StandardTypesPlugin`](/graphile-build/default-plugins/#standardtypesplugin) registers the builtin scalars,
+    or the PostGraphile plugins register the types found in the introspection results.
+5.  The schema is constructed using `newWithHooks(GraphQLSchema, …)`, where the `query`, `mutation` and `subscription` root operations
+    are configured by the respective [default plugins](/graphile-build/default-plugins)
+    and [other schema options](https://github.com/graphql/graphql-js/blob/v14.5.6/src/type/schema.js#L318-L324) can be adjusted.
```
[the remark this answers is not in the document]
Comment (ab-pm) [reply src/pages/graphile-build/hooks.md]: Problem is that the online documentation at https://graphql.org/graphql-js/type/#graphqlschema is either out of date or does deliberately omit things.  `query` and `mutation` are already handled by the default plugins, so to understand what this hook might be useful for one needs to look at the "expert" configuration options.

File: src/pages/graphile-build/hooks.md (hunk 4 above)
Comment (ab-pm) [reply src/pages/graphile-build/hooks.md]: I found it very useful to understand how the hook system works - of course ideally documentation will be good enough so that people don't have to read the source any more. Should I rather link to https://www.graphile.org/graphile-build/build-object/#newwithhookstype-spec-scope and put the details there?

File: src/pages/graphile-build/all-hooks.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (benjie) [reply src/pages/graphile-build/all-hooks.md]: Many people will only scan this section; I'd rather that it said about "adding fields" and didn't say about removing them - that way they can wonder about why removing isn't there and find that elsewhere with the full explanation. Removing fields is not an intended behaviour; it works but I would strongly advise against it.

File: src/pages/graphile-build/all-hooks.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (benjie) [reply src/pages/graphile-build/all-hooks.md]: Let's go with just the word `built` then :+1:

File: src/pages/graphile-build/all-hooks.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (benjie) [reply src/pages/graphile-build/all-hooks.md]: I've never deliberately broken a tag once it's been in the wild for more than a couple hours. Okay lets stick with tags.

File: src/pages/graphile-build/hooks.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (benjie) [reply src/pages/graphile-build/hooks.md]: Fair. I think it deliberately omits things. Also it's out of date. :+1:

File: src/pages/graphile-build/hooks.md (hunk 4 above)
Comment (benjie) [reply src/pages/graphile-build/hooks.md]: It's a philosophical thing; I think the docs should only call out to the underlying source code when that's critically important in what they're doing - for example writing inflectors. The rest of the time, the documentation should be standalone.

Adding embellishments about how it works internally is totally fine _in a footnote_ but it shouldn't be in the main body of the text. People shouldn't feel that they have to read the source in order to know how to use it. Also the source should evolve over time, but the docs should be stable - the docs outline the intended behaviour and interface, whereas the code outlines the _actual_ behaviour/interface, which is a moving target.

File: src/components/Layout/index.scss (hunk 2 above)
Comment (benjie) [reply src/components/Layout/index.scss]: OK

File: src/pages/graphile-build/all-hooks.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (ab-pm) [reply src/pages/graphile-build/all-hooks.md]: OK

File: src/pages/graphile-build/hooks.md (hunk 4 above)
Comment (ab-pm) [reply src/pages/graphile-build/hooks.md]: Is there a syntax for footnotes in the used markdown engine?

Comment (benjie) [reply src/pages/graphile-build/hooks.md]: Apparently this syntax should work; but I've not tried it:

```md
This is normal body copy.[^also] It includes a couple footnotes.[^thing]

[^also]:
  This is a footnote.

[^thing]:
  This is another footnote.
```

Comment (benjie) [reply src/pages/graphile-build/hooks.md]: We're using gatsby; so you can search "gatsby footnote" and maybe experiment with some of those plugins if you're so inclined :+1:

**Comment by ab-pm** [comment]
Thanks for merging!

> Remove localhost:8000 links

Oops. I thought I had looked out for them. Good you added a check.

> Move GraphQLSchema back to the top since it is called first

Is it really? I thought the hooks are mostly called during `init`, which precedes `GraphQLSchema`.

**Comment by benjie** [comment]
That's the order it's in (after `init`); I meant "first" as in "before `GraphQLObjectType`" :)

I didn't do any of the footnotes in the end due to lack of time.
